File: students/gilbertoanayam17/homework-4/src/banking_system.py
"""SecureBank: sistema bajo prueba para Homework 4.

Puerto uno a uno de ``src/bankingSystem.js``. Los montos se manejan como
numeros en dolares y toda operacion aritmetica pasa por ``round2`` para evitar
el arrastre de error de punto flotante (4999.99 + 0.01 da 5000.000000000001).
"""

import math

ACCOUNT_RULES = {
    "Savings": {
        "minimum_balance": 100,
        "monthly_fee": 5,
        "fee_waiver": 1000,
        "daily_limit": 2000,
    },
    "Checking": {
        "minimum_balance": 0,
        "monthly_fee": 10,
        "fee_waiver": 5000,
        "daily_limit": 5000,
    },
    "Premium": {
        "minimum_balance": 10000,
        "monthly_fee": 0,
        "fee_waiver": 0,
        "daily_limit": 50000,
    },
}

ACTIVE = "Active"
FROZEN = "Frozen"
SUSPENDED = "Suspended"
CLOSED = "Closed"

ERRORS = {
    "INVALID_ACCOUNT_TYPE": "Invalid account type",
    "INVALID_AMOUNT": "Amount must be a valid number",
    "AMOUNT_NOT_POSITIVE": "Amount must be positive",
    "BELOW_MINIMUM_AMOUNT": "Amount is below the $0.01 minimum",
    "ACCOUNT_FROZEN": "Account is frozen",
    "ACCOUNT_CLOSED": "Account is closed",
    "EXCEEDS_DAILY_LIMIT": "Exceeds daily limit",
    "INSUFFICIENT_FUNDS": "Insufficient funds",
    "INVALID_PAYEE": "Invalid payee",
    "INVALID_DATE_RANGE": "Invalid date range",
    "INVALID_TRANSITION": "Invalid state transition",
}

REGISTERED_PAYEES = ["ELECTRIC_CO", "WATER_UTILITY", "CREDIT_CARD_VISA"]

MINIMUM_AMOUNT = 0.01
DEFAULT_DATE = "2026-01-15"


def round2(value):
    """Redondea a centavos para que el punto flotante no arrastre error."""
    return round(value, 2)


def is_number(value):
    """True si el valor es un numero finito (los booleanos no cuentan)."""
    if isinstance(value, bool) or not isinstance(value, (int, float)):
        return False
    return math.isfinite(value)


# La cuenta guarda 8 atributos del dominio y el limite de pylint son 7. Juntarlos
# en un diccionario solo para bajar el numero haria mas dificiles de leer las
# pruebas de estado.
class BankAccount:  # pylint: disable=too-many-instance-attributes
    """Cuenta de SecureBank con estados, limites diarios y cobro mensual."""

    def __init__(self, account_type, initial_balance, opened_on=DEFAULT_DATE):
        if account_type not in ACCOUNT_RULES:
            raise ValueError(ERRORS["INVALID_ACCOUNT_TYPE"])
        if not is_number(initial_balance):
            raise ValueError(ERRORS["INVALID_AMOUNT"])
        if initial_balance < 0:
            raise ValueError(ERRORS["AMOUNT_NOT_POSITIVE"])

        self.account_type = account_type
        self.rules = ACCOUNT_RULES[account_type]
        self.balance = round2(initial_balance)
        self.daily_transfer_total = 0
        self.current_date = opened_on
        self.transactions = []
        self.scheduled_payments = []
        self.state = (
            SUSPENDED if self.balance < self.rules["minimum_balance"] else ACTIVE
        )

    def get_daily_limit(self):
        """Limite diario de transferencia segun el tipo de cuenta."""
        return self.rules["daily_limit"]

    def get_minimum_balance(self):
        """Saldo minimo exigido por el tipo de cuenta."""
        return self.rules["minimum_balance"]

    def set_current_date(self, date):
        """Fija la fecha del sistema (reloj inyectado para pruebas deterministas)."""
        self.current_date = date

    def reset_daily_limit(self):
        """Reinicia el acumulado diario de transferencias (corte de medianoche)."""
        self.daily_transfer_total = 0

    def sync_suspension(self):
        """Mantiene el invariante: saldo < minimo <=> Suspended."""
        if self.state in (FROZEN, CLOSED):
            return
        self.state = (
            SUSPENDED if self.balance < self.rules["minimum_balance"] else ACTIVE
        )

    def accepts_transactions(self):
        """Una cuenta congelada o cerrada no acepta movimientos."""
        return self.state in (ACTIVE, SUSPENDED)

    def blocking_state_error(self):
        """Error correspondiente al estado que bloquea la operacion."""
        if self.state == CLOSED:
            return ERRORS["ACCOUNT_CLOSED"]
        return ERRORS["ACCOUNT_FROZEN"]

    def validate_amount(self, amount):
        """Devuelve el mensaje de error del monto, o None si es valido."""
        if not is_number(amount):
            return ERRORS["INVALID_AMOUNT"]
        if amount <= 0:
            return ERRORS["AMOUNT_NOT_POSITIVE"]
        if amount < MINIMUM_AMOUNT:
            return ERRORS["BELOW_MINIMUM_AMOUNT"]
        return None

    def record(self, kind, amount, description):
        """Agrega un movimiento al historial."""
        self.transactions.append(
            {
                "date": self.current_date,
                "type": kind,
                "amount": amount,
                "description": description,
            }
        )

    def fail(self, error):
        """Resultado de una operacion rechazada."""
        return {
            "success": False,
            "error": error,
            "balance": self.balance,
            "state": self.state,
        }

    def succeed(self, **extra):
        """Resultado de una operacion aceptada."""
        result = {"success": True, "balance": self.balance, "state": self.state}
        if self.state == SUSPENDED:
            result["warning"] = "Balance is below the minimum for this account type"
        result.update(extra)
        return result

    def transfer(self, amount, destination="EXTERNAL"):
        """Transfiere dinero fuera de la cuenta aplicando todas las validaciones."""
        if not self.accepts_transactions():
            return self.fail(self.blocking_state_error())
        amount_error = self.validate_amount(amount)
        if amount_error:
            return self.fail(amount_error)

        value = round2(amount)
        if round2(self.daily_transfer_total + value) > self.rules["daily_limit"]:
            return self.fail(ERRORS["EXCEEDS_DAILY_LIMIT"])
        if value > self.balance:
            return self.fail(ERRORS["INSUFFICIENT_FUNDS"])

        self.balance = round2(self.balance - value)
        self.daily_transfer_total = round2(self.daily_transfer_total + value)
        self.record("TRANSFER", value, f"Transfer to {destination}")
        self.sync_suspension()
        return self.succeed()

    def deposit(self, amount):
        """Abona dinero y reevalua la suspension por saldo minimo."""
        if not self.accepts_transactions():
            return self.fail(self.blocking_state_error())
        amount_error = self.validate_amount(amount)
        if amount_error:
            return self.fail(amount_error)

        value = round2(amount)
        self.balance = round2(self.balance + value)
        self.record("DEPOSIT", value, "Deposit")
        self.sync_suspension()
        return self.succeed()

    def pay_bill(self, payee, amount, payment_date=None):
        """Paga o agenda un servicio a un beneficiario registrado."""
        if not self.accepts_transactions():
            return self.fail(self.blocking_state_error())
        if not isinstance(payee, str) or payee not in REGISTERED_PAYEES:
            return self.fail(ERRORS["INVALID_PAYEE"])
        amount_error = self.validate_amount(amount)
        if amount_error:
            return self.fail(amount_error)

        value = round2(amount)
        if value > self.balance:
            return self.fail(ERRORS["INSUFFICIENT_FUNDS"])

        due_date = payment_date or self.current_date
        if due_date > self.current_date:
            self.scheduled_payments.append(
                {"payee": payee, "amount": value, "due_date": due_date}
            )
            return self.succeed(scheduled=True)

        self.balance = round2(self.balance - value)
        self.record("BILL_PAYMENT", value, f"Bill payment to {payee}")
        self.sync_suspension()
        return self.succeed(scheduled=False)

    def apply_monthly_fee(self):
        """Cobro mensual del primero de mes, con exencion por saldo."""
        if self.state == CLOSED:
            return self.fail(ERRORS["ACCOUNT_CLOSED"])

        fee = self.rules["monthly_fee"]
        if fee == 0 or self.balance > self.rules["fee_waiver"]:
            return self.succeed(fee_charged=0, waived=True)
        if self.balance < fee:
            self.state = SUSPENDED
            return self.fail(ERRORS["INSUFFICIENT_FUNDS"])

        self.balance = round2(self.balance - fee)
        self.record("FEE", fee, "Monthly maintenance fee")
        self.sync_suspension()
        return self.succeed(fee_charged=fee, waived=False)

    def freeze(self):
        """Active -> Frozen por peticion del cliente o deteccion de fraude."""
        if self.state != ACTIVE:
            return self.fail(ERRORS["INVALID_TRANSITION"])
        self.state = FROZEN
        return self.succeed()

    def unfreeze(self):
        """Frozen -> Active, reevaluando el saldo minimo."""
        if self.state != FROZEN:
            return self.fail(ERRORS["INVALID_TRANSITION"])
        self.state = ACTIVE
        self.sync_suspension()
        return self.succeed()

    def close(self):
        """Cualquier estado -> Closed. Una cuenta cerrada no se reabre."""
        if self.state == CLOSED:
            return self.fail(ERRORS["ACCOUNT_CLOSED"])
        self.state = CLOSED
        return self.succeed(final_statement=self.balance)

    def get_transaction_history(self, date_from=None, date_to=None):
        """Historial filtrado por rango de fechas ISO (inclusivo)."""
        if date_from is not None and date_to is not None and date_from > date_to:
            return {
                "success": False,
                "error": ERRORS["INVALID_DATE_RANGE"],
                "transactions": [],
            }
        transactions = [
            entry
            for entry in self.transactions
            if (date_from is None or entry["date"] >= date_from)
            and (date_to is None or entry["date"] <= date_to)
        ]
        return {"success": True, "transactions": transactions}

    def export_history_to_csv(self, date_from=None, date_to=None):
        """Exporta el historial filtrado a CSV."""
        history = self.get_transaction_history(date_from, date_to)
        if not history["success"]:
            return history
        rows = [
            f"{entry['date']},{entry['type']},{entry['amount']:.2f},{entry['description']}"
            for entry in history["transactions"]
        ]
        return {
            "success": True,
            "csv": "\n".join(["date,type,amount,description"] + rows),
        }

File: students/gilbertoanayam17/homework-4/src/bankingSystem.js
/**
 * SecureBank - sistema bajo prueba para Homework 4.
 *
 * Los montos se manejan como numeros en dolares. Toda operacion aritmetica pasa
 * por round2() para evitar el arrastre de error de punto flotante (por ejemplo,
 * 4999.99 + 0.01 da 5000.000000000001 sin redondear).
 */

// Se usa un Map y no un objeto para buscar las reglas por un tipo de cuenta que
// viene de fuera: indexar un objeto con una clave variable es un riesgo de
// object injection y el linter de seguridad lo marca.
const ACCOUNT_RULES = new Map([
  [
    "Savings",
    { minimumBalance: 100, monthlyFee: 5, feeWaiver: 1000, dailyLimit: 2000 },
  ],
  [
    "Checking",
    { minimumBalance: 0, monthlyFee: 10, feeWaiver: 5000, dailyLimit: 5000 },
  ],
  [
    "Premium",
    {
      minimumBalance: 10000,
      monthlyFee: 0,
      feeWaiver: 0,
      dailyLimit: 50000,
    },
  ],
]);

const STATES = {
  ACTIVE: "Active",
  FROZEN: "Frozen",
  SUSPENDED: "Suspended",
  CLOSED: "Closed",
};

const ERRORS = {
  INVALID_ACCOUNT_TYPE: "Invalid account type",
  INVALID_AMOUNT: "Amount must be a valid number",
  AMOUNT_NOT_POSITIVE: "Amount must be positive",
  BELOW_MINIMUM_AMOUNT: "Amount is below the $0.01 minimum",
  ACCOUNT_FROZEN: "Account is frozen",
  ACCOUNT_CLOSED: "Account is closed",
  EXCEEDS_DAILY_LIMIT: "Exceeds daily limit",
  INSUFFICIENT_FUNDS: "Insufficient funds",
  INVALID_PAYEE: "Invalid payee",
  INVALID_DATE_RANGE: "Invalid date range",
  INVALID_TRANSITION: "Invalid state transition",
};

const REGISTERED_PAYEES = ["ELECTRIC_CO", "WATER_UTILITY", "CREDIT_CARD_VISA"];

const MINIMUM_AMOUNT = 0.01;
const DEFAULT_DATE = "2026-01-15";

/** Redondea a centavos para que el punto flotante no arrastre error. */
function round2(value) {
  return Math.round(value * 100) / 100;
}

class BankAccount {
  constructor(accountType, initialBalance, openedOn = DEFAULT_DATE) {
    if (!ACCOUNT_RULES.has(accountType)) {
      throw new Error(ERRORS.INVALID_ACCOUNT_TYPE);
    }
    if (
      typeof initialBalance !== "number" ||
      !Number.isFinite(initialBalance)
    ) {
      throw new Error(ERRORS.INVALID_AMOUNT);
    }
    if (initialBalance < 0) {
      throw new Error(ERRORS.AMOUNT_NOT_POSITIVE);
    }

    this.accountType = accountType;
    this.rules = ACCOUNT_RULES.get(accountType);
    this.balance = round2(initialBalance);
    this.dailyTransferTotal = 0;
    this.currentDate = openedOn;
    this.transactions = [];
    this.scheduledPayments = [];
    this.state =
      this.balance < this.rules.minimumBalance
        ? STATES.SUSPENDED
        : STATES.ACTIVE;
  }

  getDailyLimit() {
    return this.rules.dailyLimit;
  }

  getMinimumBalance() {
    return this.rules.minimumBalance;
  }

  setCurrentDate(date) {
    this.currentDate = date;
  }

  /** Reinicia el acumulado diario de transferencias (corte de medianoche). */
  resetDailyLimit() {
    this.dailyTransferTotal = 0;
  }

  /**
   * Reevalua el estado contra el saldo minimo.
   * Mantiene el invariante: saldo < minimo <=> Suspended.
   */
  syncSuspension() {
    if (this.state === STATES.FROZEN || this.state === STATES.CLOSED) {
      return;
    }
    this.state =
      this.balance < this.rules.minimumBalance
        ? STATES.SUSPENDED
        : STATES.ACTIVE;
  }

  /** Una cuenta congelada o cerrada no acepta movimientos. */
  acceptsTransactions() {
    return this.state === STATES.ACTIVE || this.state === STATES.SUSPENDED;
  }

  blockingStateError() {
    if (this.state === STATES.CLOSED) {
      return ERRORS.ACCOUNT_CLOSED;
    }
    return ERRORS.ACCOUNT_FROZEN;
  }

  /** Valida un monto de entrada; devuelve el error o null si es valido. */
  validateAmount(amount) {
    if (typeof amount !== "number" || !Number.isFinite(amount)) {
      return ERRORS.INVALID_AMOUNT;
    }
    if (amount <= 0) {
      return ERRORS.AMOUNT_NOT_POSITIVE;
    }
    if (amount < MINIMUM_AMOUNT) {
      return ERRORS.BELOW_MINIMUM_AMOUNT;
    }
    return null;
  }

  record(type, amount, description) {
    this.transactions.push({
      date: this.currentDate,
      type,
      amount,
      description,
    });
  }

  fail(error) {
    return { success: false, error, balance: this.balance, state: this.state };
  }

  succeed(extra = {}) {
    const result = { success: true, balance: this.balance, state: this.state };
    if (this.state === STATES.SUSPENDED) {
      result.warning = "Balance is below the minimum for this account type";
    }
    return Object.assign(result, extra);
  }

  transfer(amount, destination = "EXTERNAL") {
    if (!this.acceptsTransactions()) {
      return this.fail(this.blockingStateError());
    }
    const amountError = this.validateAmount(amount);
    if (amountError) {
      return this.fail(amountError);
    }

    const value = round2(amount);
    if (round2(this.dailyTransferTotal + value) > this.rules.dailyLimit) {
      return this.fail(ERRORS.EXCEEDS_DAILY_LIMIT);
    }
    if (value > this.balance) {
      return this.fail(ERRORS.INSUFFICIENT_FUNDS);
    }

    this.balance = round2(this.balance - value);
    this.dailyTransferTotal = round2(this.dailyTransferTotal + value);
    this.record("TRANSFER", value, `Transfer to ${destination}`);
    this.syncSuspension();
    return this.succeed();
  }

  deposit(amount) {
    if (!this.acceptsTransactions()) {
      return this.fail(this.blockingStateError());
    }
    const amountError = this.validateAmount(amount);
    if (amountError) {
      return this.fail(amountError);
    }

    const value = round2(amount);
    this.balance = round2(this.balance + value);
    this.record("DEPOSIT", value, "Deposit");
    this.syncSuspension();
    return this.succeed();
  }

  payBill(payee, amount, paymentDate = null) {
    if (!this.acceptsTransactions()) {
      return this.fail(this.blockingStateError());
    }
    if (typeof payee !== "string" || !REGISTERED_PAYEES.includes(payee)) {
      return this.fail(ERRORS.INVALID_PAYEE);
    }
    const amountError = this.validateAmount(amount);
    if (amountError) {
      return this.fail(amountError);
    }

    const value = round2(amount);
    if (value > this.balance) {
      return this.fail(ERRORS.INSUFFICIENT_FUNDS);
    }

    const dueDate = paymentDate || this.currentDate;
    if (dueDate > this.currentDate) {
      this.scheduledPayments.push({ payee, amount: value, dueDate });
      return this.succeed({ scheduled: true });
    }

    this.balance = round2(this.balance - value);
    this.record("BILL_PAYMENT", value, `Bill payment to ${payee}`);
    this.syncSuspension();
    return this.succeed({ scheduled: false });
  }

  /** Cobro mensual del 1o de mes, con exencion por saldo. */
  applyMonthlyFee() {
    if (this.state === STATES.CLOSED) {
      return this.fail(ERRORS.ACCOUNT_CLOSED);
    }

    const fee = this.rules.monthlyFee;
    if (fee === 0 || this.balance > this.rules.feeWaiver) {
      return this.succeed({ feeCharged: 0, waived: true });
    }
    if (this.balance < fee) {
      this.state = STATES.SUSPENDED;
      return this.fail(ERRORS.INSUFFICIENT_FUNDS);
    }

    this.balance = round2(this.balance - fee);
    this.record("FEE", fee, "Monthly maintenance fee");
    this.syncSuspension();
    return this.succeed({ feeCharged: fee, waived: false });
  }

  freeze() {
    if (this.state !== STATES.ACTIVE) {
      return this.fail(ERRORS.INVALID_TRANSITION);
    }
    this.state = STATES.FROZEN;
    return this.succeed();
  }

  unfreeze() {
    if (this.state !== STATES.FROZEN) {
      return this.fail(ERRORS.INVALID_TRANSITION);
    }
    this.state = STATES.ACTIVE;
    this.syncSuspension();
    return this.succeed();
  }

  close() {
    if (this.state === STATES.CLOSED) {
      return this.fail(ERRORS.ACCOUNT_CLOSED);
    }
    this.state = STATES.CLOSED;
    return this.succeed({ finalStatement: this.balance });
  }

  getTransactionHistory(from = null, to = null) {
    if (from !== null && to !== null && from > to) {
      return {
        success: false,
        error: ERRORS.INVALID_DATE_RANGE,
        transactions: [],
      };
    }
    const transactions = this.transactions.filter(
      (entry) =>
        (from === null || entry.date >= from) &&
        (to === null || entry.date <= to),
    );
    return { success: true, transactions };
  }

  exportHistoryToCsv(from = null, to = null) {
    const history = this.getTransactionHistory(from, to);
    if (!history.success) {
      return history;
    }
    const rows = history.transactions.map(
      (entry) =>
        `${entry.date},${entry.type},${entry.amount.toFixed(2)},${
          entry.description
        }`,
    );
    return {
      success: true,
      csv: ["date,type,amount,description", ...rows].join("\n"),
    };
  }
}

module.exports = {
  BankAccount,
  ACCOUNT_RULES,
  STATES,
  ERRORS,
  REGISTERED_PAYEES,
  MINIMUM_AMOUNT,
  round2,
};
